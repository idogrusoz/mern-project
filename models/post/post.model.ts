import { model } from "mongoose";
import PostSchema from "./post.schema";
import { IPostModel, PostDocument } from "./post.types";

export const PostModel: IPostModel = model<PostDocument, IPostModel>("Post", PostSchema);
