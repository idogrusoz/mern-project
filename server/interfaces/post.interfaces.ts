export interface IPostInterface {
    author: PostAuthor;
    textContent: string;
    style: PostStyle;
    likes: string[];
    createdAt?: Date;
    updatedAt?: Date;
}

export interface PostStyle {
    fontFamily: string;
    fontSize: number;
    fontWeight: number;
    backgroundColor: string;
    textAlign: string;
    color: string;
}

export interface PostAuthor {
    user_id: string;
    userName: string;
    image: string;
}

export interface BasePostDocument extends IPostInterface {
    _id: string;
}
