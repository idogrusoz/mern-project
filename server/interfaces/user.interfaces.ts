export interface BaseUser {
    firstName: string;
    lastName: string;
    displayName: string;
    email: string;
    userName: string;
    image?: string;
}

export interface UserInterface extends BaseUser {
    password: string;
}

export interface ResponseUser extends BaseUser {
    _id: string;
}

export interface SignInUser {
    email: string;
    password: string;
}

export interface SearchedUser {
    _id: string;
    userName: string;
    image: string;
    displayName: string;
}
