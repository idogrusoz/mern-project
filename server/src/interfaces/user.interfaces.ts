export interface UserInterface {
    firstName: string;
    lastName: string;
    displayName: string;
    email: string;
    password: string;
}

export interface ServiceResponse<T> {
    statusCode: number;
    error: boolean;
    message?: string;
    data?: T;
}
