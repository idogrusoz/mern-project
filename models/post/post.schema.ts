import { PostDocument } from "./post.types";
import { Schema } from "mongoose";

const PostSchema = new Schema({
    author: {
        user_id: {
            type: String,
            indexes: true,
            required: true,
        },
        userName: {
            type: String,
        },
        image: {
            type: String,
        },
    },
    textContent: {
        type: String,
    },
    style: {
        fontFamily: {
            type: String,
        },
        fontSize: {
            type: Number,
        },
        fontWeight: {
            type: Number,
        },
        backgroundColor: {
            type: String,
        },
        textAlign: {
            type: String,
        },
        color: {
            type: String,
        },
    },
    likes: [{ type: String, uniqueItems: true }],

    createdAt: {
        type: Date,
    },
    updatedAt: {
        type: Date,
    },
});

PostSchema.statics.findByLikes = function (userId: string) {
    var Post = this;
    return Post.find({ likes: userId });
};

PostSchema.pre<PostDocument>("save", function (this: PostDocument, next) {
    var post = this;
    if (!post.createdAt) {
        post.createdAt = new Date();
        return next();
    }
    if (!post.isModified("likes")) {
        post.updatedAt = new Date();
        return next();
    }
});

export default PostSchema;
