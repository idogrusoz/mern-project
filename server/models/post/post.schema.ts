import { PostDocument } from "./post.types";
import { Schema } from "mongoose";

const PostSchema = new Schema({
    author: {
        user_id: {
            type: String,
            indexes: true,
            required: true,
        },
        userName: {
            type: String,
        },
        image: {
            type: String,
        },
    },
    textContent: {
        type: String,
    },
    style: {
        fontFamily: {
            type: String,
        },
        fontSize: {
            type: Number,
        },
        fontWeight: {
            type: Number,
        },
        backgroundColor: {
            type: String,
        },
        textAlign: {
            type: String,
        },
        color: {
            type: String,
        },
    },
    likes: {
        type: "array",
        items: { type: "string", uniqueItems: true },
    },
    createdAt: {
        type: Date,
    },
    updatedAt: {
        type: Date,
    },
});

PostSchema.statics.findByLikes = function (userId: string) {
    var Post = this;
    return Post.find({ likes: userId });
};

PostSchema.pre<PostDocument>("save", function (this: PostDocument, next) {
    var post = this;
    if (post.isModified()) {
        post.updatedAt = new Date();
        return next();
    }
    post.createdAt = new Date();
    return next();
});

export default PostSchema;
