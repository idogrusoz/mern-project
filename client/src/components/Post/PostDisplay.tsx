import { Avatar, Card, CardActions, CardContent, CardHeader, IconButton, makeStyles } from "@material-ui/core";
import React, { FunctionComponent, useContext, useEffect, useState } from "react";
import { BasePostDocument } from "../../../../interfaces/post.interfaces";
import FavoriteIcon from "@material-ui/icons/Favorite";
import FavoriteBorderIcon from "@material-ui/icons/FavoriteBorder";
import PostCanvas from "./PostCanvas";
import { AuthContext } from "../Auth/AuthContext";
import api from "../../api";
import { ProfileContext } from "../Profile/ProfileContext";
import { useHistory } from "react-router-dom";

export type PostDisplayProps = {
    post: BasePostDocument;
    fetch: () => Promise<void>;
};

const useStyles = makeStyles((theme) => ({
    card: {
        maxWidth: 640,
    },
    icon: {
        color: theme.palette.secondary.dark,
    },
    header: {
        cursor: "pointer",
    },
}));

const PostDisplay: FunctionComponent<PostDisplayProps> = ({ post, fetch }) => {
    const classes = useStyles();
    const { user } = useContext(AuthContext);
    const { profileOwner } = useContext(ProfileContext);
    const { style, textContent, likes } = post;
    const { backgroundColor, color, fontFamily, fontSize, fontWeight, textAlign } = style;
    const [liked, setLiked] = useState<boolean | null>(null);
    const { push } = useHistory();
    const handleLike = async () => {
        try {
            const response = await api.put(`/posts/${post._id}/like`, { like: !liked });
            if (!response.data.error) {
                setLiked(response.data.data.likes.indexOf(user!._id) > -1);
                if (!profileOwner) {
                    fetch();
                }
            }
        } catch (error) {
            throw error;
        }
    };

    useEffect(() => {
        const like = likes.indexOf(user!._id) > -1;
        setLiked(like);
    }, [likes, user, post._id]);

    return (
        <Card className={classes.card}>
            <CardHeader
                avatar={<Avatar aria-label="recipe" src={post.author.image} />}
                title={post.author.userName}
                subheader={
                    new Date(post.createdAt!).toDateString() + " " + new Date(post.createdAt!).toLocaleTimeString()
                }
                onClick={() => push(`${post.author.user_id}`)}
                className={classes.header}
            />
            <CardContent>
                <PostCanvas
                    backgroundColor={backgroundColor}
                    color={color}
                    fontFamily={fontFamily}
                    fontSize={fontSize}
                    fontWeight={fontWeight}
                    textAlign={textAlign as "left" | "right" | "justify" | "center"}
                    textContent={textContent}
                />
            </CardContent>
            <CardActions disableSpacing>
                <IconButton aria-label="like" onClick={handleLike} className={classes.icon}>
                    {liked ? <FavoriteIcon /> : <FavoriteBorderIcon />}
                </IconButton>
            </CardActions>
        </Card>
    );
};

export default PostDisplay;
