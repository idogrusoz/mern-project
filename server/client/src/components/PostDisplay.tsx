import { Avatar, Card, CardContent, CardHeader, makeStyles } from "@material-ui/core";
import React, { FunctionComponent } from "react";
import { IPostInterface } from "../../../interfaces/post.interfaces";
import PostCanvas from "./PostCanvas";

type PostDisplayProps = {
    post: IPostInterface;
};

const useStyles = makeStyles({
    card: {
        maxWidth: 640,
    },
});

const PostDisplay: FunctionComponent<PostDisplayProps> = ({ post }) => {
    const classes = useStyles();
    const { style, textContent } = post;
    const { backgroundColor, color, fontFamily, fontSize, fontWeight, textAlign } = style;
    return (
        <Card className={classes.card}>
            <CardHeader
                avatar={<Avatar aria-label="recipe" src={post.author.image} />}
                title={post.author.userName}
                subheader={
                    new Date(post.updatedAt!).toDateString() + " " + new Date(post.updatedAt!).toLocaleTimeString()
                }
            />
            <CardContent>
                <PostCanvas
                    backgroundColor={backgroundColor}
                    color={color}
                    fontFamily={fontFamily}
                    fontSize={fontSize}
                    fontWeight={fontWeight}
                    textAlign={textAlign as "left" | "right" | "justify" | "center"}
                    textContent={textContent}
                />
            </CardContent>
        </Card>
    );
};

export default PostDisplay;
