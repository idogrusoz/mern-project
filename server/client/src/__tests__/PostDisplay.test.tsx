import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import PostDisplay from "../components/PostDisplay";
import { IPostInterface } from "../../../interfaces/post.interfaces";

describe("PostDisplay test", () => {
    const props: { post: IPostInterface } = {
        post: {
            likes: [],
            textContent: "",
            author: {
                user_id: "userId",
                image: "",
                userName: "userName",
            },
            style: {
                fontFamily: "Roboto",
                color: "#ffffff",
                backgroundColor: "#ffffff",
                fontWeight: 400,
                fontSize: 40,
                textAlign: "left" as "left" | "right" | "justify" | "center",
            },
        },
    };

    const wrapper = shallow(<PostDisplay {...props} />);
    it("renders without crash", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
