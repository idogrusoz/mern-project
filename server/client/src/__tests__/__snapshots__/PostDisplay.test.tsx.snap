// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDisplay test renders without crash 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-card-1"
>
  <WithStyles(ForwardRef(CardHeader))
    avatar={
      <ForwardRef(WithStyles)
        aria-label="recipe"
      />
    }
    subheader="Invalid Date Invalid Date"
    title="userId"
  />
  <WithStyles(ForwardRef(CardContent))>
    <PostCanvas
      backgroundColor="#ffffff"
      color="#ffffff"
      fontFamily="Roboto"
      fontSize={40}
      fontWeight={400}
      textAlign="left"
      textContent=""
    />
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;
